File: core/shared/go/credentiallease/lease_test.go
package credentiallease_test

import (
	"bytes"
	"testing"
	"time"

	"cdsoft.com.cn/VastPlan/core/shared/go/credentiallease"
	"cdsoft.com.cn/VastPlan/core/shared/go/pluginconfig"
)

func TestLeaseRoundTripIsBoundAndOneUse(t *testing.T) {
	ref := pluginconfig.ManagedCredentialRef{Handle: "credential://managed/abc", Scope: "tenant", Owner: "plugin.database", Purpose: "database.connection", Version: 2}
	request, recipient, err := credentiallease.NewRequest(ref)
	if err != nil {
		t.Fatal(err)
	}
	now := time.Now().UTC()
	claims := credentiallease.Claims{TenantID: "tenant-a", Audience: "node-a", Ref: ref}
	secret := []byte("database-password")
	envelope, err := credentiallease.Seal(request, claims, secret, now, 10*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	if bytes.Contains([]byte(envelope.Ciphertext), secret) {
		t.Fatal("lease ciphertext 不得包含明文")
	}
	material, err := recipient.Open(envelope, claims, now.Add(time.Second))
	if err != nil || !bytes.Equal(material, secret) {
		t.Fatalf("解封失败 material=%q err=%v", material, err)
	}
	for index := range material {
		material[index] = 0
	}
	if _, err := recipient.Open(envelope, claims, now.Add(time.Second)); err == nil {
		t.Fatal("同一 recipient 不得重复消费")
	}
}

func TestLeaseRejectsTamperingAudienceAndExpiry(t *testing.T) {
	ref := pluginconfig.ManagedCredentialRef{Handle: "credential://managed/abc", Scope: "tenant", Owner: "plugin.database", Purpose: "database.connection", Version: 1}
	now := time.Now().UTC()
	claims := credentiallease.Claims{TenantID: "tenant-a", Audience: "node-a", Ref: ref}

	request, recipient, _ := credentiallease.NewRequest(ref)
	envelope, _ := credentiallease.Seal(request, claims, []byte("secret"), now, time.Second)
	envelope.Audience = "node-b"
	if _, err := recipient.Open(envelope, credentiallease.Claims{TenantID: "tenant-a", Audience: "node-b", Ref: ref}, now); err == nil {
		t.Fatal("篡改 AAD 必须失败")
	}

	request, recipient, _ = credentiallease.NewRequest(ref)
	envelope, _ = credentiallease.Seal(request, claims, []byte("secret"), now, time.Second)
	if _, err := recipient.Open(envelope, claims, now.Add(2*time.Second)); err == nil {
		t.Fatal("过期 lease 必须失败")
	}
}

func TestLeaseRejectsMalformedManagedReference(t *testing.T) {
	_, _, err := credentiallease.NewRequest(pluginconfig.ManagedCredentialRef{Handle: "not-managed", Scope: "tenant", Owner: "plugin", Purpose: "database", Version: 1})
	if err == nil {
		t.Fatal("非 managed handle 必须拒绝")
	}
}

File: extensions/sdk/node/credential-lease/test/lease.test.mjs
import assert from "node:assert/strict";
import {
  createCipheriv,
  createHmac,
  createPublicKey,
  diffieHellman,
  generateKeyPairSync,
  randomBytes,
} from "node:crypto";
import test from "node:test";

import { MaterialLeaseClient } from "../src/index.mjs";

const prefix = Buffer.from("302a300506032b656e032100", "hex");
const ref = {
  handle: "credential://managed/oidc",
  scope: "tenant",
  owner: "cn.vastplan.foundation.security.authentication.provider.oidc",
  purpose: "oidc.client-secret",
  version: 1,
};

function seal(request, now) {
  const recipientRaw = Buffer.from(request.recipientPublicKey, "base64url");
  const recipient = createPublicKey({
    key: Buffer.concat([prefix, recipientRaw]),
    format: "der",
    type: "spki",
  });
  const sender = generateKeyPairSync("x25519");
  const shared = diffieHellman({
    privateKey: sender.privateKey,
    publicKey: recipient,
  });
  const salt = randomBytes(32);
  const prk = createHmac("sha256", salt).update(shared).digest();
  const key = createHmac("sha256", prk)
    .update("vastplan/material-lease/v1")
    .update(Buffer.from([1]))
    .digest();
  const nonce = randomBytes(12);
  const senderDer = sender.publicKey.export({ format: "der", type: "spki" });
  const envelope = {
    version: 1,
    leaseId: "lease-test",
    tenantId: "acme",
    audience: "node-runtime.test",
    ref,
    issuedAtUnixMs: now,
    expiresAtUnixMs: now + 15_000,
    senderPublicKey: senderDer
      .subarray(senderDer.length - 32)
      .toString("base64url"),
    salt: salt.toString("base64url"),
    nonce: nonce.toString("base64url"),
  };
  const aad = Buffer.from(
    JSON.stringify({
      version: envelope.version,
      leaseId: envelope.leaseId,
      tenantId: envelope.tenantId,
      audience: envelope.audience,
      ref: envelope.ref,
      issuedAtUnixMs: envelope.issuedAtUnixMs,
      expiresAtUnixMs: envelope.expiresAtUnixMs,
    }),
  );
  const cipher = createCipheriv("aes-256-gcm", key, nonce);
  cipher.setAAD(aad);
  const encrypted = Buffer.concat([
    cipher.update("confidential-secret"),
    cipher.final(),
    cipher.getAuthTag(),
  ]);
  envelope.ciphertext = encrypted.toString("base64url");
  return envelope;
}

test("Node Material Lease opens a bound envelope and clears the callback buffer", async () => {
  const now = Date.UTC(2026, 6, 23);
  let material;
  const plugin = {
    async call(_target, context, payload) {
      assert.equal(context.tenant_id, "acme");
      const request = JSON.parse(Buffer.from(payload).toString());
      return {
        result: { status: "STATUS_OK" },
        payload: Buffer.from(JSON.stringify(seal(request, now))),
      };
    },
  };
  const client = new MaterialLeaseClient(plugin, {
    audience: "node-runtime.test",
    now: () => now,
  });
  const value = await client.withMaterial(
    ref,
    "acme",
    undefined,
    async (secret) => {
      material = secret;
      return secret.toString();
    },
  );
  assert.equal(value, "confidential-secret");
  assert.ok(material.every((byte) => byte === 0));
});

test("Node Material Lease rejects audience tampering", async () => {
  const now = Date.UTC(2026, 6, 23);
  const plugin = {
    async call(_target, _context, payload) {
      const envelope = seal(JSON.parse(Buffer.from(payload).toString()), now);
      envelope.audience = "attacker";
      return {
        result: { status: "STATUS_OK" },
        payload: Buffer.from(JSON.stringify(envelope)),
      };
    },
  };
  const client = new MaterialLeaseClient(plugin, {
    audience: "node-runtime.test",
    now: () => now,
  });
  await assert.rejects(
    () => client.withMaterial(ref, "acme", undefined, async () => {}),
    /claims/,
  );
});
